atomix/cluster/src/test/java/io/atomix/raft/protocol/ControllableRaftServerProtocol.java

```diff
@@ -0,0 +1,423 @@
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Function;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Controllable raft server protocol. The messages are delivered only when explicitly instructed and
+ * messages can be dropped.
+ */
+public class ControllableRaftServerProtocol implements RaftServerProtocol {
+
+  private Function<JoinRequest, CompletableFuture<JoinResponse>> joinHandler;
+  private Function<LeaveRequest, CompletableFuture<LeaveResponse>> leaveHandler;
+  private Function<ConfigureRequest, CompletableFuture<ConfigureResponse>> configureHandler;
+  private Function<ReconfigureRequest, CompletableFuture<ReconfigureResponse>> reconfigureHandler;
+  private Function<InstallRequest, CompletableFuture<InstallResponse>> installHandler;
+  private Function<TransferRequest, CompletableFuture<TransferResponse>> transferHandler;
+  private Function<PollRequest, CompletableFuture<PollResponse>> pollHandler;
+  private Function<VoteRequest, CompletableFuture<VoteResponse>> voteHandler;
+  private Function<AppendRequest, CompletableFuture<AppendResponse>> appendHandler;
+  private final Set<MemberId> partitions = Sets.newCopyOnWriteArraySet();
+  private final Map<MemberId, ControllableRaftServerProtocol> servers;
+  // Incoming messages to each member
+  private final Map<MemberId, Queue<Tuple<Runnable, CompletableFuture>>> messageQueue;
+  private final MemberId localMemberId;
+
+  public ControllableRaftServerProtocol(
+      final MemberId memberId,
+      final Map<MemberId, ControllableRaftServerProtocol> servers,
+      final Map<MemberId, Queue<Tuple<Runnable, CompletableFuture>>> messageQueue) {
+    this.servers = servers;
+    this.messageQueue = messageQueue;
+    localMemberId = memberId;
+    messageQueue.put(memberId, new LinkedList<>());
+    servers.put(memberId, this);
+  }
+
+  public void receiveNextMessage() {
+    final var rcvQueue = messageQueue.get(localMemberId);
+    if (!rcvQueue.isEmpty()) {
+      rcvQueue.poll().getLeft().run();
+    }
+  }
+
+  public void receiveAll() {
+    final var rcvQueue = messageQueue.get(localMemberId);
+    while (!rcvQueue.isEmpty()) {
+      rcvQueue.poll().getLeft().run();
+    }
+  }
+
+  // drop next message from the incoming queue
+  public void dropNextMessage() {
+    final var nextMessage =
+        messageQueue.computeIfAbsent(localMemberId, t -> new LinkedList<>()).poll();
+    if (nextMessage != null) {
+      Optional.ofNullable(nextMessage.getRight())
+          .ifPresent(
+              f -> {
+                LoggerFactory.getLogger("TEST:")
+                    .info("Dropped a message to {}", localMemberId.id());
+                // RaftServers excepts exceptions from the messaging layer to detect timeouts
+                f.completeExceptionally(new TimeoutException());
```

Comment:
If we just drop the message we should also receive at some point an timeout right? Is it worth to test this?

Reply:
For elections, there are timeouts inbuilt in raft roles. But for appendEntries from leader to followers, it relies on timeout from the communication layer.

Reply:
ah ok weird